Blog post about custom errors (#135)
ethereum/solidity-blog: _posts/2021-04-20-custom-errors.md

---
layout: post
published: true
title:  Custom Errors in Solidity
date:   2021-04-20
author: Solidity Team
category: Explainers
---

[Solidity v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4) provides a convenient

> and gas-efficient way to explain to user why an operation failed. Errors can be defined inside and

Reviewer: I'd say either ``explain to an user`` or ``explain to users``.
```suggestion
and gas-efficient way to explain to users why an operation failed. Errors can be defined inside and
```

outside of contracts (including interfaces and libraries).

## Example

The following contract shows an example usage of an Error:

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.4;

contract VendingMachine {
    address payable owner = payable(msg.sender);
    error Unauthorized();
    function withdraw() public {
        if (msg.sender != owner)
            revert Unauthorized();

        owner.transfer(address(this).balance);
    }
    // ...
}
```

> The syntax of Errors are similar to that of

Reviewer: Why capitalize Errors here? I'd just say ``errors``. And is syntax not singular? (not sure, but for me it sounds more right)

```suggestion
The syntax of errors is similar to that of
```

> [Events](https://docs.soliditylang.org/en/latest/contracts.html#events). They have to be used

Reviewer: ```suggestion
[events](https://docs.soliditylang.org/en/latest/contracts.html#events). They have to be used
```

together with the [revert
statement](https://docs.soliditylang.org/en/latest/control-structures.html#revert-statement) which
causes all changes in the current call to be reverted and passes the error data back to the caller.

> ### Errors in depth

Reviewer: ```suggestion
### Errors in Depth
```

In the above example, `revert Unauthorized();` is equivalent to the following Yul code:

```
   let free_mem_ptr := mload(64)
   mstore(free_mem_ptr, 0x82b4290000000000000000000000000000000000000000000000000000000000)
   revert(free_mem_ptr, 4)
```

This is the same as the ABI encoding of a function call with the name `Unauthorized()`. Here
`0x82b42900` is the 'selector' for `Unauthorized()`. In contrast, using a revert string, i.e.,
`revert("Unauthorized");` leads to the following Yul code:

```
    let free_mem_ptr := mload(64)
    mstore(free_mem_ptr, 0x08c379a000000000000000000000000000000000000000000000000000000000)
    mstore(add(free_mem_ptr, 4), 32)
    mstore(add(free_mem_ptr, 36), 12)
    mstore(add(free_mem_ptr, 68), "Unauthorized")
    revert(free_mem_ptr, 100)
```

Here `0x08c379a` is the 'selector' of `Error(string)`. One can see that custom Errors can decrease
both deploy (always) and runtime gas costs (only when the revert condition is met).

## Errors with Parameters

> It is also possible to have errors that take Parameters. For example,

Reviewer: ```suggestion
It is also possible to have errors that take parameters. For example,
```

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.4;

/// Insufficient balance for transfer. Needed `required` but only
/// `available` available.
/// @param available balance available.
/// @param required requested amount to transfer.
error InsufficientBalance(uint256 available, uint256 required);

contract TestToken {
    mapping(address => uint) balance;
    function transfer(address to, uint256 amount) public {
        if (amount > balance[msg.sender])
            // Error call using named parameters. Equivalent to
            // InsufficientBalance(balance[msg.sender], amount)
            revert InsufficientBalance({
                available: balance[msg.sender],
                required: amount
            });
        balance[msg.sender] -= amount;
        balance[to] += amount;
    }
    // ...
}
```

The error data would be encoded identically as the ABI encoding for function calls, i.e.,
``abi.encodeWithSignature("InsufficientBalance(uint256,uint256)", balance[msg.sender], amount)``.

> We hope that frameworks will provide support for Errors. The following is an example on decoding

Reviewer: ?
```suggestion
We hope that frameworks will provide support for custom errors. The following is an example on decoding
```

            "type": "uint256"
        },
        {
            "internalType": "uint256",
            "name": "required",
            "type": "uint256"
        }
    ],
    "name": "InsufficientBalance",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
}];

const interface = new ethers.utils.Interface(abi);
// Error data obtained after calling `TestToken.transfer(...)`.
const error_data = "0xcf47918100000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000100000000";
interface.decodeFunctionData(interface.functions['InsufficientBalance(uint256,uint256)'], error_data);
// [
//   BigNumber { _hex: '0x0100', _isBigNumber: true },
//   BigNumber { _hex: '0x0100000000', _isBigNumber: true },
//   available: BigNumber { _hex: '0x0100', _isBigNumber: true },
//   required: BigNumber { _hex: '0x0100000000', _isBigNumber: true }
// ]
```

The compiler includes all errors that a contract can emit in the contract's
[ABI-JSON](https://docs.soliditylang.org/en/latest/abi-spec.html?#json). Note that this will not
include errors emitted through external calls. Similarly, developers can provide

> [NatSpec](https://docs.soliditylang.org/en/latest/natspec-format.html) documentation for Errors

Reviewer: ```suggestion
[NatSpec](https://docs.soliditylang.org/en/latest/natspec-format.html) documentation for errors
```

which would then be part of the user and developer documentation.

Note that one can never trust the error data. The error data by default bubbles up through the chain
of external calls, which means that a contract may receive an error not defined in any of the
contracts it calls directly. Furthermore, any contract can fake any error by returning data that
matches an error signature, even if the error is not defined anywhere.

> Currently, there is no convenient way to catch errors in Solidity, but this is planned.

Reviewer: Do we have an issue or anything that we can link to or can we make the "this is planned" more detailed or sound a bit nicer?

Author: I couldn't find any issue. I'll make a new one.

---
layout: post
published: true
title:  Custom Errors in Solidity

> date:   2021-04-20

Author: A note to self to update the date before merging.

Reviewer 2: And the file name

---
layout: post
published: true
title:  Custom Errors in Solidity
date:   2021-04-20
author: Solidity Team
category: Explainers
---

> [Solidity v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4) provides a convenient
> and gas-efficient way to explain to users why an operation failed. Errors can be defined inside and
> outside of contracts (including interfaces and libraries).

Reviewer 2: ```suggestion
Starting from [Solidity v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4), there is a convenient
and gas-efficient way to explain to users why an operation failed through the use of custom errors.
Until now, you could already use strings to give more information about failures (``revert("Insufficient funds.");``),
but they are rather expensive and it is difficult to use dynamic information in them.

Custom errors are defined using the ``error`` statement, which can be used inside and outside of contracts (including interfaces and libraries).
```

## Example

The following contract shows an example usage of an Error:

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.4;

contract VendingMachine {
    address payable owner = payable(msg.sender);

>     error Unauthorized();

Reviewer 2: Maybe we should move it outside of the contract so that it stands out more?

    function withdraw() public {
        if (msg.sender != owner)
            revert Unauthorized();

        owner.transfer(address(this).balance);
    }
    // ...
}
```

The syntax of errors is similar to that of
[events](https://docs.soliditylang.org/en/latest/contracts.html#events). They have to be used
together with the [revert
statement](https://docs.soliditylang.org/en/latest/control-structures.html#revert-statement) which
causes all changes in the current call to be reverted and passes the error data back to the caller.

> 

Reviewer 2: Maybe add: Using errors together with ``require`` is not yet supported (see below)

and then further down explain that ``require(c, "abc")`` should be translated into ``if (!c) revert Abc();``

revert(free_mem_ptr, 100)
```

Here `0x08c379a` is the 'selector' of `Error(string)`. In this comparison, one can see that custom
errors decrease both deploy and runtime gas costs. Note that runtime gas cost is only relevant when
the revert condition is met.

## Errors with Parameters

It is also possible to have errors that take parameters. For example,

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.4;

/// Insufficient balance for transfer. Needed `required` but only
/// `available` available.
/// @param available balance available.
/// @param required requested amount to transfer.
error InsufficientBalance(uint256 available, uint256 required);

contract TestToken {
    mapping(address => uint) balance;
    function transfer(address to, uint256 amount) public {
        if (amount > balance[msg.sender])
            // Error call using named parameters. Equivalent to
            // InsufficientBalance(balance[msg.sender], amount)
            revert InsufficientBalance({
                available: balance[msg.sender],
                required: amount
            });
        balance[msg.sender] -= amount;
        balance[to] += amount;
    }
    // ...
}
```

> The error data would be encoded identically as the ABI encoding for function calls, i.e.,

Reviewer 2: Why not use this instead of inline assembly above?

``abi.encodeWithSignature("InsufficientBalance(uint256,uint256)", balance[msg.sender], amount)``.

We hope that frameworks will provide support for custom errors. The following is an example on
decoding error data using [ethers.js](https://docs.ethers.io/v5/):

```javascript
import { ethers } from "ethers";

// As a workaround, we have a function with the
// same name and parameters as the error in the abi.

> const abi = [{
>     "inputs": [{
>             "internalType": "uint256",
>             "name": "available",
>             "type": "uint256"
>         },
>         {
>             "internalType": "uint256",
>             "name": "required",
>             "type": "uint256"
>         }
>     ],
>     "name": "InsufficientBalance",
>     "outputs": [],
>     "stateMutability": "nonpayable",
>     "type": "function"
> }];

Reviewer 2: One reason  why ethers is far superior to web3.js:

```suggestion
const abi = ["function InsufficientBalance(uint256, uint256)"];
```

Author: This would only return

```
[
  BigNumber { _hex: '0x0100', _isBigNumber: true },
  BigNumber { _hex: '0x0100000000', _isBigNumber: true }
]
```

With the full abi-json, it's nice to see the parameters `available` and `required` mapped to values.

Reviewer 2: Does it help if you use `function InsufficientBalance(uint256 available, uint256 required)`?

Author: Wow. That actually works. I'll change the abi.

const interface = new ethers.utils.Interface(abi);
const error_data = "0xcf47918100000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000100000000";
interface.decodeFunctionData(
    interface.functions['InsufficientBalance(uint256,uint256)'],
    error_data
);
// [
//   BigNumber { _hex: '0x0100', _isBigNumber: true },
//   BigNumber { _hex: '0x0100000000', _isBigNumber: true },
//   available: BigNumber { _hex: '0x0100', _isBigNumber: true },
//   required: BigNumber { _hex: '0x0100000000', _isBigNumber: true }
// ]
```

The compiler includes all errors that a contract can emit in the contract's
[ABI-JSON](https://docs.soliditylang.org/en/latest/abi-spec.html?#json). Note that this will not

> include errors emitted through external calls. Similarly, developers can provide

Reviewer 2: ```suggestion
include errors forwarded through external calls. Similarly, developers can provide
```
?

[NatSpec](https://docs.soliditylang.org/en/latest/natspec-format.html) documentation for errors

> which would then be part of the user and developer documentation.

Reviewer 2: ```suggestion
which would then be part of the user and developer documentation and can explain the error in much more detail at no cost.
```

> Note that one can never trust the error data. The error data by default bubbles up through the chain
> of external calls, which means that a contract may receive an error not defined in any of the
> contracts it calls directly. Furthermore, any contract can fake any error by returning data that
> matches an error signature, even if the error is not defined anywhere.

Reviewer 2: ```suggestion
Please be careful when using error data since its origin is not tracked. The error data by default bubbles up through the chain
of external calls, which means that a contract may forward an error not defined in any of the
contracts it calls directly. Furthermore, any contract can fake any error by returning data that
matches an error signature, even if the error is not defined anywhere.
```

decoding error data using [ethers.js](https://docs.ethers.io/v5/):

```javascript
import { ethers } from "ethers";

// As a workaround, we have a function with the
// same name and parameters as the error in the abi.
const abi = [{
    "inputs": [{
            "internalType": "uint256",
            "name": "available",
            "type": "uint256"
        },
        {
            "internalType": "uint256",
            "name": "required",
            "type": "uint256"
        }
    ],
    "name": "InsufficientBalance",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
}];

const interface = new ethers.utils.Interface(abi);
const error_data = "0xcf47918100000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000100000000";
interface.decodeFunctionData(
    interface.functions['InsufficientBalance(uint256,uint256)'],
    error_data
);
// [
//   BigNumber { _hex: '0x0100', _isBigNumber: true },
//   BigNumber { _hex: '0x0100000000', _isBigNumber: true },
//   available: BigNumber { _hex: '0x0100', _isBigNumber: true },
//   required: BigNumber { _hex: '0x0100000000', _isBigNumber: true }

> // ]

Author: @ricmoo This is currently how we suggested dealing with errors. Do you have anything to add here?

---
layout: post
published: true
title:  Custom Errors in Solidity
date:   2021-04-20
author: Solidity Team
category: Explainers
---

> [Solidity v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4) provides a convenient and gas-efficient way to explain to user why an operation failed. They can be defined inside and outside of contracts (including interfaces and libraries). 

Reviewer 2: Please avoid long lines. What does "They" refer to?

> The syntax of Errors are similar to that of

Reviewer 2: Maybe rather put this after the example? It could be confusing to mention events here, because only the syntax is similar, not the semantics.

[Events](https://docs.soliditylang.org/en/latest/contracts.html#events). They have to be used
together with the [revert
statement](https://docs.soliditylang.org/en/latest/control-structures.html#revert-statement) which
causes all changes in the current call to be reverted and passes the error data back to the caller.

The following contract shows an example usage of an Error:

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.4;

contract VendingMachine {
    address payable owner = payable(msg.sender);
    error Unauthorized();
    function withdraw() public {
        if (msg.sender != owner)
            revert Unauthorized();

        owner.transfer(address(this).balance);
    }
    // ...
}
```

> In the above example, `revert Unauthorized();` is equivalent to the following Yul code:

Reviewer 2: Isn't this going a bit too deep?

Reviewer 2: Maybe we should rather highlight the cost difference compared to `revert("Unauthorized")`?

Author: I wanted to include something that makes it clear that solidity isn't randomly assigning numbers to errors. Perhaps I can rephrase this slightly better, but still include the Yul code. 

I'll also highlight the cost difference.

Reviewer 2: If it's not too complicated, maybe we should really just show how to decode it using ethers.js - in that way the encoding should be clear, shouldn't it?

Reviewer 2: People will stop reading when they see "depth" and inline assembly. We can use this, but then it should be put at the very end of the blog post.
